src/csm_testing/tests/check_workflow_template/__main__.py

```diff
@@ -0,0 +1,259 @@
+
+    # Check the status of containers in the pod
+    for container_status in pod.status.container_statuses:
+        if container_status.name == "main":  # Assuming main container is named "main"
+            if container_status.state.terminated:
+                return True
+    return False
+
+def check_and_kill_pod(workflow_name, namespace):
+    v1 = client.CoreV1Api()
+    
+    # Get the list of pods
+    pods = v1.list_namespaced_pod(namespace=namespace, label_selector=f"workflows.argoproj.io/workflow={workflow_name}").items
+    
+    for pod in pods:
+        pod_name = pod.metadata.name
+        print(f"INFO: Checking pod: {pod_name}")
+        
+        # Check if the main container has finished
+        if is_main_container_finished(pod_name, namespace):
+            print(f"INFO: Main container finished in pod {pod_name}. Deleting pod.")
+            v1.delete_namespaced_pod(name=pod_name, namespace=namespace)
+            print(f"INFO: Pod {pod_name} deleted.")
+            return True
+        else:
+            print(f"INFO: Main container still running in pod {pod_name}.")
+
+def update_image_version_in_template(workflow_template_str,old_image_prefix,old_version, new_version):
+    # Find the image string and replace its version
+    updated_workflow_template_str = workflow_template_str.replace(
+        old_image_prefix + old_version, 
+        old_image_prefix + new_version
+    )
+    return updated_workflow_template_str
+
+def wait_for_workflow_to_succeed(namespace, workflow_name, timeout=6000, interval=20):
+    """
+    Polls the workflow status until it succeeds or the timeout is reached.
+    """
+    start_time = time.time()
+    while time.time() - start_time < timeout:
+        try:
+            # Get the workflow status
+            command = [
+            "kubectl", "get", "workflow", workflow_name, "-n", namespace, "-o", "jsonpath={.status.phase}"
+            ]
+            result = subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True)
+            status=result.stdout.strip()
+            
+            # Check the status of the workflow
+            print(f"Workflow {workflow_name} status: {status}")
+            
+            if status == "Succeeded":
+                print(f"INFO: Workflow {workflow_name} succeeded.")
+                return True
+            elif status in ["Failed", "Error"]:
+                print(f"ERROR: Workflow {workflow_name} failed with status {status}.")
+                return False
+            elif status in ["Running"]:
+                print("INFO; Check if main is done")
+                if check_and_kill_pod(workflow_name, namespace):
+                    print(f"WARNING: Workflow {workflow_name} is running but main container is done.")
+                    return True
+        except ApiException as e:
+            print(f"ERROR: Exception when fetching workflow status: {e}")
```

Comment:
That message makes it look like the try statement should be around the `kubectl` command, since that is what is retrieving the status.

But this `except` block is catching `ApiException`, which seems like it would only be raised by the call to `check_and_kill_pod`. 

Either way, the try block should not enclose such a large block of code. It should surround the part that is expected to raise the exception, it should catch the appropriate exception, and the message should be relevant to what was being attempted.

Reply:
I moved the except block , now it will catch subprocess error for kubectl command.
```
try:
            # Get the workflow status
            command = [
                "kubectl",
                "get",
                "workflow",
                workflow_name,
                "-n",
                namespace,
                "-o",
                "jsonpath={.status.phase}",
            ]
            result = subprocess.run(
                command,
                check=True,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                universal_newlines=True,
            )
        except subprocess.CalledProcessError as err:
            print(f"ERROR: Exception when fetching workflow status: {err}")
            return False
```